Add ACME base URL parameter

edewata opened the pull request:
By default the ACME directory will return ACME service URLs
with the same hostname that the client uses to access the
directory. If the hostname is load-balanced, the client might
get redirected to different servers, which could trigger other
issues.

A new parameter has been added into `engine.conf` to override
the base URL of ACME services. This mechanism can be used to
pin the client to the current server.

A COPR build is available here:
https://copr.fedorainfracloud.org/coprs/edewata/acme/builds/

For example, configure this in `/etc/pki/pki-tomcat/acme/engine.conf` and restart the server:
```
baseURL=https://server1.example.com/acme
```
Then access the directory using the load-balanced hostname:
```
$ curl -s -k https://ipa-ca.example.com/acme/directory | python -m json.tool
```
Suppose the directory is served by server1.example.com, the client will get:
```
{
    "meta": {
        "caaIdentities": [
            "example.com"
        ],
        "externalAccountRequired": false,
        "termsOfService": "https://www.example.com/acme/tos.pdf",
        "website": "https://www.example.com"
    },
    "newAccount": "https://server1.example.com/acme/new-account",
    "newNonce": "https://server1.example.com/acme/new-nonce",
    "newOrder": "https://server1.example.com/acme/new-order",
    "revokeCert": "https://server1.example.com/acme/revoke-cert"
}
```
Presumably the client will be able to complete certificate enrollment using server1.example.com.

rcritten commented:
It isn't clear what this does. Is the intention to return the "real" CA hostname via the Location header to the client?

Isn't the url value in the JWS supposed to match the responding host?

edewata commented:
@rcritten I've added some explanation above. I'm not sure whether this is fully compliant with the RFC, but I did some basic testing with certbot and it seems to be working fine. Could you test with a real IPA cluster?

frasertweedale commented:
@rcritten @edewata :

RFC 8555 doesn't say anything about directory object including URLs at a different hostname(s).  So, it is legal per spec.  But... it might confuse clients.  We need to test it with a variety of clients.

Also need to make sure that the ipa-pki-proxy config lets the requests for the server hostname through (i.e. initial access of directory object may be for `ipa-ca.$domain/acme/directory`, with subsequent requests for `$host,$domain`/acme/...`.  Off top of my head, I think that will Just Work with the current configuration.  But again, need to verify that before merging this change.

frasertweedale reviewed (commented):
One nit, but overall the implementation seems fine.  I have not tested yet.  @rcritten will or someone on the IPA team have cycles to test this change with IPA?

edewata commented on base/acme/src/main/java/org/dogtagpki/acme/server/ACMEDirectoryService.java:
```diff
@@ -36,21 +37,30 @@ public Response getDirectory() throws Exception {
+        URI baseURL = engine.getBaseURL();
+
+        UriBuilder uriBuilder;
+        if (baseURL != null) {
```
Actually, since the `baseURL` is supposed to point to an actual location instead of just an identifier, I changed the type from `URI` into `URL`. See `ACMEEngineConfig.fromProperties()`. The `URL` constructor there will reject empty/invalid base URL in `engine.conf`.

edewata commented:
Thanks for the feedback. I've updated the patch and the build.

frasertweedale commented:
Thinking about a test scenario for this, it should be something like:

1. Installl IPA server with CA.
2. Ensure `ipa-ca.$domain` DNS name points at it.
3. Enable ACME.
4. Update `/etc/pki/pki-tomcat/acme/engine.conf` to include `baseURL=https://${SERVER_FQDN}/acme` and restart Dogtag.
4. On another system ("client"), enrol as a freeipa client (or manually add the IPA CA to ca-trust).
5. On client, using $ACME_CLIENT program, create account and request certificate
6. Install IPA replica (with CA).
7. Decommission original IPA server.
8. Ensure `ipa-ca.$domain` points to new IPA server.
9. Update replica's `/etc/pki/pki-tomcat/acme/engine.conf` to include `baseURL=https://${REPLICA_FQDN}/acme` and restart Dogtag.
10. Use $ACME_CLIENT to renew the certificate.  (This may mean you have to modify the profile or use a custom profile with a very short lifetime, because some clients including certbot refuse to renew certificates are are too far from expiry).

If that all succeeds, *for all clients we care about*, then we have confirmed that this is a valid approach to ACME load balancing.

frasertweedale commented:
I think the account key rollover (which we do not yet implement IIRC) does not seem to fit well with this LB approach.
https://tools.ietf.org/html/rfc8555#section-7.3.5
```
   To create this request object, the client first constructs a
   keyChange object describing the account to be updated and its account
   key:

   account (required, string):  The URL for the account being modified.
      The content of this field MUST be the exact string provided in the
      Location header field in response to the newAccount request that
      created the account.
```
However, we can look up the account by the old key, and validate that the supplied `account` URL "matches" it.  So it is not a showstopper, up to the possibility of clients being "too clever" and inspecting URLs.

frasertweedale commented:
Account deactivation is more problematic, when the account URL is bound to a retired server.

https://tools.ietf.org/html/rfc8555#section-7.3.6

```
7.3.6.  Account Deactivation

   A client can deactivate an account by posting a signed update to the
   account URL with a status field of "deactivated".
```

frasertweedale commented:
@edewata so, to address the issues discussed above, perhaps we need *two* config: one a statically configured `baseURLShared`, which provides the load-balanced name, and `baseURLServer` which provides the server-specific name.

Objects that need a stable URL over a long timespan (especially accounts, per examples above) should always be referenced through `baseURLShared`, and other objects dealt with over a short timespan (orders, authorizations, etc) can use the server-specific base URL.

Other observations:

* the revocation operation does not refer to certificate URLs; so no problems here (thankfully)
* authz deactivation *does* use the authz URL.  So this operation may fail after topology changes.
* authz reuse is OK (if we ever want to implement it), because we construct URLs from the challenge-ID "on the fly".  For a new order, the re-used authzs will appear with a URL based on the "current server".

edewata commented:
I've rebased this PR against PR #3436. I have not implemented the latest suggestion yet, let me do some investigation.

edewata commented:
@frasertweedale I've rebased this PR against PR #3442 and added a CI test to verify server switchover using the `baseURL` parameter. It looks like the client is fine, so it's not necessary to add another parameter. Could you take a look? Thanks.

frasertweedale commented:
> @frasertweedale I've rebased this PR against PR #3442 and added a CI test to verify server switchover using the `baseURL` parameter. It looks like the client is fine, so it's not necessary to add another parameter. Could you take a look? Thanks.

The operations that I am concerned about are:

- Account key rollover
- Account deactivation
- Authorization deactivation

We do not implement these yet.  So we cannot test it.  Besides, I'm not sure whether certbot or any clients even supports those operations either.

On that basis, I have no objection to merging this.  If we implement any of the those features in the future and it turns out to be needed, we can add it then (there could be upgrade consequences).

Oh, and thank you for implementing the tests!

edewata commented:
@frasertweedale Thanks! The switchover test runs `certbot unregister` as well, which I suppose does an account deactivation. FYI, @rcritten has tested this PR using multiple IPA servers: https://github.com/freeipa/freeipa/pull/5464

@cipherboy Are we OK to merge this PR?

cipherboy reviewed (approved):
Looks good. It sounds like we need to do a bit more testing in the future, but it should be possible to manually work around some of these issues by editing the ACME account config file (`/etc/letsencrypt/accounts/<host>/directory/<account>/regr.json`).

My 2c.

edewata commented:
@rcritten @frasertweedale @cipherboy Thanks! Merging..
